[Jai Lulla] iP

Jai2501 opened the pull request:
```
  ____      _        _                         _   _               ____ _           _   _           _    
 |  _ \ ___| |_ _ __(_) _____   _____ _ __    | |_| |__   ___     / ___| |__   __ _| |_| |__   ___ | |_  
 | |_) / _ \ __| '__| |/ _ \ \ / / _ \ '__|   | __| '_ \ / _ \   | |   | '_ \ / _` | __| '_ \ / _ \| __| 
 |  _ <  __/ |_| |  | |  __/\ V /  __/ |      | |_| | | |  __/   | |___| | | | (_| | |_| |_) | (_) | |_  
 |_| \_\___|\__|_|  |_|\___| \_/ \___|_|       \__|_| |_|\___|    \____|_| |_|\__,_|\__|_.__/ \___/ \__| 
                                                                                                                                                                                                                                                                                                                                         
```
> The Dog themed Chatbot that gets all your tasks done! (and bites you if you don't) :dog2:

What's Retriever good at?
* Retrieving your data
* _~~Biting~~_ **Requesting** you to complete your tasks
* Playing with your life

Okay, on a serious note:
1. It's Fast
2. It can let you search up for tasks
3. It also lets you delete the tasks that you completed _~~(we know you didn't)~~_

## Features Implemented
- [x] Managing Todos
- [x] Managing Deadlines
- [x] Managing Events
- [x] Marking tasks as done
- [x] Deleting tasks
- [ ] GUI

## Code

This amazing Chatbot can be found **[here](https://github.com/Jai2501/ip)**.

This is the `main` method:
```Java
public static void main(String[] args) {
    // Calling the run() method to start the Chatbot.
    new Retriever("tasksList.txt").run();
}
```
:bulb: You can edit the file path in this line `new Retriever("filePath").run();`, to store the tasks at a place you wish to!

Commit:
Add Gradle support

Jai2501 committed:
Add Level 1 Features to the Chatbot

The Chatbot should be able to interact with the user.

To achieve that, the first step done is to echo back what
the user enters.

The Chatbot would shutdown when the
user inputs "bye"(Case Sensitivity is taken care of).

Jai2501 committed:
Add Level 2 Features to the chatbot

It would be good if we can store the user entered
inputs, and the user could view those any
time he would like to do so.

Add an List<String> to store the various
user inputs. Register some keywords
including "bye" and "list" to terminate
the session with the chatbot and to view
the items present in the list respectively.

List<String> is used as we can store
almost infinite inputs versus how arrays
would restrict to a certain number that has
to be determined at the start of the program.

Jai2501 committed:
Add Level 3 features to the Chatbot

To make the chatbot more functional, and to help
you through your daily chores, this chatbot can now
keep a track of various tasks you have.

A new class called Task is added to help track the
task status(whether done or not). Also, the ability to
mark tasks as done is added.

Jai2501 committed:
Add Level 4 features to the Chatbot

There are various types of tasks including but
not limited to events, deadlines and todos.
The chatbot should be able to serve the end user
better and be more functional.

Extracting the common behaviour into a super class,
called "Task" allows to use polymorphism when dealing
with various types of tasks.

Jai2501 committed:
Add A-TextUiTesting Level Features to Chatbot

A Chatbot should provide appropriate response. This
implies, we shall do enough testing of the Chatbot.

Performed Testing on the Chatbot and updated
.gitignore file.

Jai2501 committed:
Add Level 5 features to the Chatbot

The Chatbot should not crash at any instance if
an invalid/unformatted input is entered by the user.
It should instead inform the user to use the right
format and perhaps show an error message as well.

Custom exceptions are made to take care of various
scenarios where a user might enter a wrong input.

Also updated input and EXPECTED text files for
testing purpose.

Jai2501 committed:
Add Level 6 Features to the Chatbot

The scenario where the user enters task details
and makes some mistake, or if the user has decided
not to do the task on that particular day, he/she can delete
the task.

Added a delete functionality to allow users to delete tasks.

Updated testing files accordingly.

Jai2501 committed:
Add Level 8 features to the Chatbot

Timing is something that the bot needs to
be aware of. Representing it merely as String
makes our life easy as developers, but in terms
of functionality, it doesn't help.

Therefore, a custom class is added to handle
date and it's format.

Jai2501 committed:
Add Level 7 features to the Chatbot

For the chatbot to be functional, it
should be able to store users previous
tasks. Even on relaunching the bot, it
shall showcase all the previous tasks
user had, whether accomplished or not.

Add a Storage class to help implement this
functionality using a text file.

Jai2501 committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add Level 7 features to the Chatbot

Jai2501 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add Level 8 features to the Chatbot

Jai2501 committed:
Alter the Task representation in text file

Since date also used '/', there were problems
in parsing the Task when read from a file.

Change '/' to '|' for Task representation in Text file.

Jai2501 committed:
Add A-MoreOOP to the Chatbot

In the Retriever chatbot, most of the code
is in the main - Retriever class.

The goal is to refactor the code to extract out
closely related code as classes to make the
code more OOP.

Add various classes including Ui, Storage,
Parser and TaskList.

Jai2501 committed:
Add A-Packages to the Chatbot

The files were all organised in one
folder. Searching for files was becoming
complex as the number of files kept on
increasing.

Add packages. Organize the files into
exception, task, and retriever packages
respectively.

To avoid naming conflicts and
to make it easier to search for various files,
packaging has been implemented.

Jai2501 committed:
Delete Files

Delete the files that have been
reorganized into various packages.

Jai2501 committed:
Add A-JUnit to the Chatbot

Testing is an integral part of programming
apps. It is done to ensure the smooth running
of a program.

Add JUnit Test classes.

Jai2501 committed:
Add A-Jar to the Chatbot

To release the app for public use.

Create a JAR file.

It can be located in:
out/artifacts/individualProject_jar/individualProject.jar

Jai2501 committed:
Add Level-9 feature to the Chatbot

When the task list grows considerably in
size, it becomes hard to look for tasks.

Add find feature.

This allows to look up for a task with
a given keyword.

Jai2501 committed:
Add A-CodingStandard to the Chatbot

Following a coding standard is necessary
as it makes the code look well formatted and
enhances readability.

Refine the code according to the
given Java coding standard.

Jai2501 committed:
Add A-JavaDoc to the Chatbot

Documentation helps others to understand
what your code is trying to do.

Add JavaDoc comments.

Jai2501 committed:
Fix runtest.sh

Testing is important. It ensures that
your program delivers what you
expect from it.

Made changes to runtest.sh file and other
files that help to test the chatbot.

Some other files have also been added and
deleted.

Jai2501 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add A-JavaDoc to the Chatbot

Jai2501 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Add A-CodingStandard to the Chatbot

# Conflicts:
#	src/main/java/retriever/Ui.java
#	src/main/java/retriever/task/Task.java

Conflicts resolved.

Jai2501 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add Level-9 feature to the Chatbot

# Conflicts:
#	src/main/java/retriever/Parser.java
#	src/main/java/retriever/Ui.java

Conflicts resolved.

Jai2501 committed:
Refine Code Style

Jai2501 committed:
Implement Ui in Storage class

Jai2501 committed:
Remove methods from Task.java

Removed some unused methods from
Task.java file.

Jai2501 committed:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d'

* commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d':
  Add Gradle support

Jai2501 committed:
Add Gradle support to the Chatbot

Added Gradle to automate various build
tasks and testing of the Chatbot project

Jai2501 committed:
Merge branch 'branch-A-Gradle'

* branch-A-Gradle:
  Add Gradle support to the Chatbot

Jai2501 committed:
Add A-CheckStyle support

Added checkStyle to detect coding style violations.

Jai2501 committed:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Add A-CheckStyle support

Jai2501 committed:
Tweak code to follow new code style

After adding checkStyle, some coding style
violations were detetcted.

Hence, these have been fixed.

Jai2501 committed:
use basic JavaFX

Leofeng10 commented on src/main/java/retriever/Parser.java:
```diff
@@ -0,0 +1,84 @@
+     * Marks the isSessionDone to true, when the user inputs
+     * the "bye" keyword, indicating, the end of session.
+     */
+    public void markSessionDone() {
+        isSessionDone = true;
+    }
+
+    /**
+     * Returns whether the user would like to continue or no.
+     *
+     * @return A boolean indicating to end the session or not.
+     */
+    public boolean isExit() {
+        return isSessionDone;
+    }
```
should this two functions' names be similar since they are working on the same variable?

Leofeng10 commented on src/main/java/retriever/Storage.java:
```diff
@@ -0,0 +1,230 @@
+                String taskDescription = tokens[2];
+
+                // Checking the type of task.
+                if (taskType.contains("D")) {
+                    Task deadlineTask = new Deadline(taskDescription, new TaskDateAndTime(tokens[3]));
+
+                    if (completionStatus.contains("1")) {
+                        deadlineTask.markAsDone();
+                    }
+
+                    taskList.add(deadlineTask);
+                }
+
+                if (taskType.contains("E")) {
+                    Task eventTask = new Event(taskDescription, new TaskDateAndTime(tokens[3]));
+
+                    if (completionStatus.contains("1")) {
+                        eventTask.markAsDone();
+                    }
+
+                    taskList.add(eventTask);
+                }
+
+                if (taskType.contains("T")) {
+                    Task todoTask = new Todo(taskDescription);
+
+                    if (completionStatus.contains("1")) {
+                        todoTask.markAsDone();
+                    }
```
is using else if here a better option than using three if statements?

Leofeng10 commented on src/main/java/retriever/task/TaskList.java:
```diff
@@ -0,0 +1,226 @@
+     */
+    public void addEventTask(String userInput) throws IllegalEventFormatException, IllegalDateFormatException {
+        // Making sure that a properly formatted event task is entered.
+        if (!userInput.toLowerCase().contains("/at")
+                || userInput.split(" ").length < 4
+                || userInput.substring(5).split(" /at ").length < 2) {
```
should be make the condition more clear here by breaking them into a few parts?

Leofeng10 reviewed (commented):
Great work done. Nice code structure. Very readable code! 👍

Jai2501 committed:
Add GUI to the Chatbot

Having a good interaction medium, that is,
a Graphical User Interface(GUI) to interact
with the chatbot is a must. It also
improves the usability of the app.

Jai2501 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add GUI to the Chatbot
  use basic JavaFX

Jai2501 committed:
Fix code style issues

Jai2501 committed:
Add Welcome Splash Screen

Jai2501 committed:
Fix Text Truncate Issues for Larger Text

Zhouyirui commented on src/main/java/retriever/task/TaskList.java:
```diff
@@ -0,0 +1,217 @@
+            taskStorage.deleteTask(taskNumber);
+            ui.printTaskDeleted(task, taskListLength());
+        } else {
+            throw new TaskNotFoundException("Sorry! The Task Number Entered Does Not Exist!");
```
Perhaps you can get either useTaskList.remove or taskStorage.deletedTask to throw the exception for array out of bounds?

Zhouyirui commented on src/main/java/retriever/task/TaskList.java:
```diff
@@ -0,0 +1,217 @@
+     * @param parsedUserInput The parsed command, entered by the user with the "find" keyword.
+     */
+    public void findTaskWithKeyword(String[] parsedUserInput) {
+        String taskDescription;
```
Perhaps declare this in line 200 so it will have a smaller scope?

Zhouyirui commented on src/main/java/retriever/task/TaskList.java:
```diff
@@ -0,0 +1,217 @@
+ */
+public class TaskList {
+    private ArrayList<Task> userTaskList;
+    private Storage taskStorage;
```
I feel TaskList doesnt have a HAS-A relationship with Storage? So perhaps it would be better if Storage was in charge reading from the database and returning a ArrayList of tasks for TaskList to initialize

Zhouyirui commented on src/main/java/retriever/Ui.java:
```diff
@@ -0,0 +1,205 @@
+     * @param taskListLength The length of the task list, after adding the task.
+     */
+    public void printTaskAdded(Task task, int taskListLength) {
+        System.out.println("-> Where's My Treat? I Added:\n\t" + task);
+        System.out.println("\nYou Owe Me " + taskListLength + " Treat(s), Master!");
```
Would it be better if you assign the string to retrieverResponse first, then System.out.println(retrieverResponse)?
```suggestion
        retrieverResponse = "Where's My Treat? I Added:\n"
                + task + "\nYou Owe Me " + taskListLength + " Treat(s), Master!";
        System.out.println(retrieverResponse");
```

Zhouyirui commented on src/main/java/retriever/Ui.java:
```diff
@@ -0,0 +1,205 @@
+ * user and the chatbot.
+ */
+public class Ui {
+    private static String retrieverResponse;
```
I like that you have this variable! Its a simple way to allow Retriever to work on both CLI and GUI 😄

Zhouyirui reviewed (commented):
Overall, OOP is well done! Intuitive and easy to understand.

Jai2501 committed:
Add quit functionality to the Chatbot

Jai2501 committed:
Update code to create a new JAR file

Jai2501 committed:
Update files to improve GUI support

A GUI was added to the Chatbot, and therefore
the CLI based U.I. became obsolete. Consequently,
there is a need to remove the dead code.

Remove various lines of code that help run the
CLI version of the Chatbot.

This would enhance the code quality.

Jai2501 committed:
Enhance UI

Improved the Welcome Splash screen
visuals.

Jai2501 committed:
Improve Code Quality

The code works at the moment; however, the code might be a bit
messy or unclear at specific instances.

The code needs to be refactored as this would improve the
readability, and also simplify various complex expressions
present.

Refactor the code as per the Code Quality guide.

The guide is being followed to refactor to adhere to the coding
standards set up in this module.

Jai2501 committed:
Add Assertions

The code works fine. However, there is a risk of it crashing due to
some assumptions that we made during writing the program to become
false during the execution.

Add Assertions at various places to verify that the assumptions made
are true.

Assertions are being adopted because they help define assumptions about
the program state so the runtime can verify them.

Jai2501 committed:
Merge pull request #1 from Jai2501/branch-A-Assertions

Add Assertions

Jai2501 committed:
Merge branch 'master' of https://github.com/Jai2501/ip

* 'master' of https://github.com/Jai2501/ip:
  Add Assertions

Jai2501 committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add Assertions

Jai2501 committed:
Merge pull request #2 from Jai2501/branch-A-CodeQuality

Improve Code Quality

Jai2501 committed:
Merge branch 'master' of https://github.com/Jai2501/ip

* 'master' of https://github.com/Jai2501/ip:
  Improve Code Quality

Jai2501 committed:
Fix code style issues

Jai2501 committed:
Add help page

The chatbot is relatively easy to use once you get the hang of it.
However, for a new user, it might be a daunting task to find out
more about the commands available, and how to use them.

Add a help page, which shows a list of commands available and
how to use them.

The most intuitive way of looking for help would be by typing help
in the text field.

Jai2501 committed:
Merge branch 'branch-C-Help'

* branch-C-Help:
  Add help page

Jai2501 committed:
Merge branch 'master' into branch-UI-Enhancements

* master:
  Add help page
  Fix code style issues
  Add Assertions
  Improve Code Quality

# Conflicts:
#	src/main/java/retriever/Ui.java

Jai2501 committed:
Update welcome splash screen to show help message

Jai2501 committed:
Add gradle.yml file

Jai2501 committed:
delete I/O redirection tests

Jai2501 committed:
Merge branch 'branch-A-CI'

* branch-A-CI:
  delete I/O redirection tests
  Add gradle.yml file

Jai2501 committed:
Add ability to view schedules

The users can view the tasks by using the list command. It allows them
to see which tasks they have completed and which they haven't.

The users need more power and flexibility to be able to view the
tasks they have to do by the end of a particular day.

Add view command that lets the users look at the tasks due on a
particular day.

The ability to specify the date helps the user to customize the
command.

Jai2501 committed:
Add comments

Jai2501 committed:
Fix code style issues

Jai2501 committed:
Merge branch 'branch-B-ViewSchedules'

* branch-B-ViewSchedules:
  Fix code style issues
  Add comments
  Add ability to view schedules

Jai2501 committed:
Refactor Code

Made subtle changes to the way date is handled and checked.

Jai2501 committed:
Update JUnit test cases

Jai2501 committed:
Tweak code to follow code style

Jai2501 committed:
Update help message to show 'view'

Jai2501 committed:
Add JUnit Tests for Ui

Jai2501 committed:
Merge branch 'master' into branch-UI-Enhancements

* master:
  Add JUnit Tests for Ui
  Update help message to show 'view'
  Tweak code to follow code style
  Update JUnit test cases
  Refactor Code
  Fix code style issues
  Add comments
  Add ability to view schedules
  delete I/O redirection tests
  Add gradle.yml file

# Conflicts:
#	src/main/java/retriever/Ui.java

Jai2501 committed:
Add background color to dialog box

Jai2501 committed:
Enhance Dialog Box Ui

It is important to showcase the messages in
a organised manner.

Tweak the GUI to match the asymmetric nature
of the conversation.

Jai2501 committed:
Merge branch 'branch-UI-Enhancements'

* branch-UI-Enhancements:
  Enhance Dialog Box Ui
  Add background color to dialog box
  Update welcome splash screen to show help message
  Enhance UI

Jai2501 committed:
Refactor Ui Class

Jai2501 committed:
Change UiTest Directory

Jai2501 committed:
Minor Change to TaskDateAndTimeTest

Jai2501 committed:
Tweak code to follow code style

Jai2501 committed:
Refactor TaskListTest

Jai2501 committed:
Add JUnit Tests for Parser

Jai2501 committed:
Tweak code to follow code style

Jai2501 committed:
Fix Ui Format for find

Jai2501 committed:
Update README.md

README.md updated to show the User Guide.

Jai2501 committed:
Update JUnit Test Case for UiTest

Jai2501 committed:
Set theme jekyll-theme-midnight

Jai2501 committed:
Update Checkstyle.xml to support constant names

Jai2501 committed:
Merge branch 'master' of https://github.com/Jai2501/ip

* 'master' of https://github.com/Jai2501/ip:
  Set theme jekyll-theme-midnight

Jai2501 committed:
Update README.md

Jai2501 committed:
Update README.md

Jai2501 committed:
Update README.md

Jai2501 committed:
Update README.md

Jai2501 committed:
Update README.md

Jai2501 committed:
Update README.md

Jai2501 committed:
Update README.md and Version Number

Jai2501 committed:
Add Ui.png

Jai2501 committed:
Update README.md

Jai2501 committed:
Merge branch 'master' of https://github.com/Jai2501/ip

* 'master' of https://github.com/Jai2501/ip:
  Update README.md

Jai2501 committed:
Update JAR link in README.md

Jai2501 committed:
Update README.md

Jai2501 committed:
Update README.md

Jai2501 committed:
Update README.md

Jai2501 committed:
Edit JavaDoc Comments in DialogBox.java

Jai2501 committed:
Update README.md

Jai2501 committed:
Update README.md

Jai2501 committed:
Update README.md

Jai2501 committed:
Add Log capability to Storage Class

To reduce clutter on chat interface and to avoid
distracting the user with error messages, these messages
should be placed elsewhere, where they are noticeable.

Add Log capabilities to the Storage class.

Any error message with regards to File operations and Storage
class would be logged to the console and stored in a file
'storageLog.log' as well.

Jai2501 committed:
Update Ui Class

Updated Ui Class to adapt the help message
to show how to check logs.

Jai2501 committed:
Merge branch 'branch-add-logger'

* branch-add-logger:
  Update Ui Class
  Add Log capability to Storage Class

Jai2501 committed:
Tweak code to follow code style

Jai2501 committed:
Update Log Messages

Jai2501 committed:
Update README.md
